core/src/code.rs

```diff
@@ -322,172 +382,178 @@ impl Code {
-                    }
-                    _ => None,
-                })
-                .ok_or(CodeError::MemoryEntryNotFound)?,
-        )
-        .map_err(|_| CodeError::InvalidStaticPageCount)?;
-
-        let exports = get_exports(&module, false)?;
-
-        if check_entries
+        let exports = get_exports(&module, config.check_exports)?;
+        if config.check_exports
             && !(exports.contains(&DispatchKind::Init) || exports.contains(&DispatchKind::Handle))
         {
             return Err(CodeError::RequiredExportFnNotFound);
         }
 
-        let code = if instrument_with_const_rules {
-            let instrumented_module =
-                gear_wasm_instrument::inject(module, &ConstantCostRules::default(), "env")
-                    .map_err(|_| CodeError::GasInjection)?;
+        if let Some(stack_limit) = config.stack_height {
+            module = wasm_instrument::inject_stack_limiter(module, stack_limit).map_err(|err| {
+                log::trace!("Failed to inject stack height limits: {err}");
+                CodeError::StackLimitInjection
+            })?;
+        }
 
-            parity_wasm::elements::serialize(instrumented_module).map_err(|_| CodeError::Encode)?
-        } else {
-            original_code.clone()
-        };
+        if let Some(mut get_gas_rules) = get_gas_rules {
+            let gas_rules = get_gas_rules(&module);
+            module = gear_wasm_instrument::inject(module, &gas_rules, "env")
+                .map_err(|_| CodeError::GasInjection)?;
+        }
+
+        let code = parity_wasm::elements::serialize(module).map_err(|err| {
+            log::trace!("Failed to encode instrumented program: {err}");
+            CodeError::Encode
+        })?;
 
         Ok(Self {
-            raw_code: original_code,
             code,
+            original_code,
             exports,
             static_pages,
-            instruction_weights_version: version,
+            instruction_weights_version: config.version,
         })
     }
 
-    /// Create the code with instrumentation, but without checks.
-    /// There is also no check for static memory pages.
-    pub fn new_raw_with_rules<R, GetRulesFn>(
+    /// Create the code by checking and instrumenting `original_code`.
+    /// Main logic of instrumentation can be represented by this example:
+    /// Let's take a code:
+    /// ```wasm
+    /// (module
+    ///    (import "env" "memory" (memory 1))
+    ///    (export "init" (func $init))
+    ///    (func $f1
+    ///       <-- f1 code -->
+    ///    )
+    ///    (func $f2
+    ///       if (i32.eqz (i32.const 0))
+    ///          <-- some code -->
+    ///       else
+    ///          <-- some code -->
+    ///       end
+    ///    )
+    ///    (func $f3
+    ///       <-- f3 code -->
+    ///    )
+    ///    (func $init
+    ///       call $f1
+    ///       call $f2
+    ///       call $f3
+    ///       <-- some code -->
+    ///    )
+    /// )
```

Comment:
It would be nice to make docs more shorter and better, for example use only two funcs.
```
/// (module
    ///    (import "env" "memory" (memory 1))
    ///    (export "init" (func $init))
    ///
    ///    (func $f1
    ///       <-- f1 code -->
    ///    )
    ///
    ///    (func $f2
    ///       if (i32.eqz (i32.const 0))
    ///          <-- some code -->
    ///       else
    ///          <-- some code -->
    ///       end
    ///    )
    ///
    ///    (func $init
    ///       call $f1
    ///       call $f2
    ///       <-- some code -->
    ///    )
    /// )
```

Reply:
done